code/control/ApprovalsDispatcher.php

```diff
@@ -0,0 +1,299 @@
+					'role' => $data['RoleTitle'],
+					'name' => $data['FullName']
+				];
+			}
+		}
+
+		return $this->getAPIResponse([
+			'approvers' => $list
+		], 200);
+	}
+
+	/**
+	 * @param \SS_HTTPRequest $request
+	 * @return \SS_HTTPResponse
+	 */
+	public function submit(SS_HTTPRequest $request) {
+		if ($request->httpMethod() !== 'POST') {
+			return $this->getAPIResponse(['message' => 'Method not allowed, requires POST'], 405);
+		}
+
+		$deployment = DNDeployment::get()->byId($request->postVar('id'));
+		$errorResponse = $this->validateDeployment($deployment);
+		if ($errorResponse instanceof \SS_HTTPResponse) {
+			return $errorResponse;
+		}
+
+		$approver = Member::get()->byId($request->postVar('approver_id'));
+		if (!$approver || !$approver->exists()) {
+			return $this->getAPIResponse(['message' => 'Invalid approver. Does not exist'], 403);
+		}
+		if (!$this->project->allowed(self::ALLOW_APPROVAL, $approver)) {
+			return $this->getAPIResponse(['message' => 'The given approver does not have permissions to approve'], 403);
+		}
+
+		try {
+			$deployment->ApproverID = $approver->ID;
+			$deployment->write();
+
+			$deployment->getMachine()->apply(DNDeployment::TR_SUBMIT);
+		} catch (\Exception $e) {
+			return $this->getAPIResponse([
+				'message' => $e->getMessage()
+			], 400);
+		}
+
+		return $this->getAPIResponse([
+			'message' => 'Deployment request has been submitted',
+			'deployment' => $this->formatter->getDeploymentData($deployment)
+		], 200);
+	}
+
+	/**
+	 * @param \SS_HTTPRequest $request
+	 * @return \SS_HTTPResponse
+	 */
+	public function cancel(SS_HTTPRequest $request) {
+		if ($request->httpMethod() !== 'POST') {
+			return $this->getAPIResponse(['message' => 'Method not allowed, requires POST'], 405);
+		}
+
+		$deployment = DNDeployment::get()->byId($request->postVar('id'));
+		$errorResponse = $this->validateDeployment($deployment);
+		if ($errorResponse instanceof \SS_HTTPResponse) {
+			return $errorResponse;
+		}
+
+		// @todo permission checking for cancelling an approval request
+		try {
+			$deployment->getMachine()->apply(DNDeployment::TR_NEW);
+		} catch (\Exception $e) {
+			return $this->getAPIResponse([
+				'message' => $e->getMessage()
+			], 400);
+		}
+
+		return $this->getAPIResponse([
+			'message' => 'Deployment request has been cancelled',
+			'deployment' => $this->formatter->getDeploymentData($deployment)
+		], 200);
+	}
+
+	/**
+	 * @param \SS_HTTPRequest $request
+	 * @return \SS_HTTPResponse
+	 */
+	public function approve(SS_HTTPRequest $request) {
+		if ($request->httpMethod() !== 'POST') {
+			return $this->getAPIResponse(['message' => 'Method not allowed, requires POST'], 405);
+		}
+
+		$deployment = DNDeployment::get()->byId($request->postVar('id'));
+		$errorResponse = $this->validateDeployment($deployment);
+		if ($errorResponse instanceof \SS_HTTPResponse) {
+			return $errorResponse;
+		}
+
+		$validStates = [DNDeployment::STATE_SUBMITTED, DNDeployment::STATE_NEW];
```

Comment:
Turns out I didn't need to do this, just check for permissions and then let the transition take care of doing it underneath (it'll throw a bad API response if the state machine errored)